src/main/kotlin/com/pubnub/api/eventengine/EffectDispatcher.kt

```diff
@@ -0,0 +1,35 @@
+package com.pubnub.api.eventengine
+
+import java.util.concurrent.ConcurrentHashMap
+
+interface ManagedEffect {
+    fun run(completionBlock: () -> Unit = {})
+    fun cancel()
+}
+
+class EffectDispatcher<T : EffectInvocation>(
+    private val effectHandlerFactory: EffectHandlerFactory<T>,
+    private val managedEffects: ConcurrentHashMap<String, ManagedEffect> = ConcurrentHashMap()
+) {
+
+    fun dispatch(effectInvocation: T) {
+        when (effectInvocation) {
+            is CancelEffectInvocation -> {
+                managedEffects.remove(effectInvocation.idToCancel)?.cancel()
+            }
+
+            is ManagedEffectInvocation -> {
+                managedEffects.remove(effectInvocation.id)?.cancel()
+                val managedEffect = effectHandlerFactory.create(effectInvocation)
+                managedEffects[effectInvocation.id] = managedEffect
+                managedEffect.run {
+                    managedEffects.remove(effectInvocation.id)
+                }
+            }
+
+            else -> {
```

Comment:
The plan was to not have any exceptional situations actually. The hierarchy should be known (is known because it's a `T`), and handled by the effect handlers. I'll try to come up with a way to have `NonManaged` and at the same time not have `else`

Reply:
ok